google-api-client/src/main/java/com/google/api/client/googleapis/services/AbstractGoogleClient.java

```diff
@@ -328,9 +447,34 @@ protected Builder(
         HttpRequestInitializer httpRequestInitializer) {
       this.transport = Preconditions.checkNotNull(transport);
       this.objectParser = objectParser;
-      setRootUrl(rootUrl);
-      setServicePath(servicePath);
+      this.rootUrl = normalizeRootUrl(rootUrl);
+      this.servicePath = normalizeServicePath(servicePath);
       this.httpRequestInitializer = httpRequestInitializer;
+      this.serviceName = parseServiceName(rootUrl);
+      this.isUserConfiguredEndpoint = !defaultEndpointRegex.matcher(this.rootUrl).matches();
+    }
+
+    /**
+     * This is intended to invoked once on the initial Builder's constructor call. This parses the
+     * rootUrl value (set from the Discovery Doc) to use for the serviceName. The serviceName is
+     * used to construct an endpoint when the user passes in a custom Universe Domain value.
+     *
+     * <p>The roolUrl from the Discovery Docs will always follow the format of
+     * https://{serviceName}(.mtls).googleapis.com/
+     */
+    private String parseServiceName(String rootUrl) {
```

Comment:
I think we can use the regex to extract the service name now.

Reply:
Sounds good, will update,